File: android/src/main/kotlin/io/sentient/android/chat/Typewriter.kt
// Typewriter — pure time→reveal engine porting webui use-typewriter-buffer.ts.
// Rates from shared tokens (mobilesdk.design.Typewriter). A Compose driver
// (rememberTypewriterText) calls typewriterTick per frame; this stays pure/testable.
package io.sentient.android.chat

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.runtime.withFrameNanos
import io.sentient.mobilesdk.design.Typewriter as Cfg

data class TypewriterState(val visibleCount: Int = 0, val pauseUntil: Double = 0.0)

private val sentencePause = Cfg.sentencePauseMs / 1000.0
private val paragraphPause = Cfg.paragraphPauseMs / 1000.0

/** Advance visibleCount toward target.size for elapsed [dt] at clock [now] (seconds). */
fun typewriterTick(
    s: TypewriterState,
    target: List<Char>,
    streamComplete: Boolean,
    dt: Double,
    now: Double,
): TypewriterState {
    val n = target.size
    if (s.visibleCount >= n) return s
    if (now < s.pauseUntil) return s
    val gap = n - s.visibleCount
    val raw = if (streamComplete) Cfg.maxRate.toDouble() else Cfg.baseRate * (1 + gap * Cfg.gapGain)
    val rate = minOf(Cfg.maxRate.toDouble(), maxOf(Cfg.minRate.toDouble(), raw))
    val advance = maxOf(1, (rate * dt).toInt())
    // CRITICAL: stop the advance AT the first sentence/paragraph boundary in the
    // window [visibleCount+1 .. candidate], so a fast tick can't skip a pause.
    val candidate = minOf(n, s.visibleCount + advance)
    val stop = firstBoundaryStop(target, s.visibleCount + 1, candidate)
    val newCount = stop ?: candidate
    val pause = boundaryPause(target, newCount)
    return s.copy(visibleCount = newCount, pauseUntil = if (pause != null) now + pause else s.pauseUntil)
}

/** First index (1-based count) in [from..through] whose just-revealed char is a boundary; else null. */
private fun firstBoundaryStop(t: List<Char>, from: Int, through: Int): Int? {
    var i = from
    while (i <= through) {
        if (boundaryPause(t, i) != null) return i
        i++
    }
    return null
}

private fun boundaryPause(t: List<Char>, upto: Int): Double? {
    if (upto < 1 || upto > t.size) return null
    val last = t[upto - 1]
    if (last == '\n' && upto >= 2 && t[upto - 2] == '\n') return paragraphPause
    if (last == '.' || last == '!' || last == '?') return sentencePause
    return null
}

/** Revealed text for a streaming bubble; full text when not streaming. No cursor. */
@Composable
fun rememberTypewriterText(content: String, streaming: Boolean): String {
    if (!streaming) return content
    val latest by rememberUpdatedState(content)
    var state by remember { mutableStateOf(TypewriterState()) }
    LaunchedEffect(Unit) {
        var last = 0.0
        while (true) {
            withFrameNanos { nanos ->
                val now = nanos / 1_000_000_000.0
                val dt = if (last == 0.0) 0.0 else now - last
                last = now
                state = typewriterTick(state, latest.toList(), streamComplete = false, dt = dt, now = now)
            }
        }
    }
    val chars = latest.toList()
    return chars.subList(0, state.visibleCount.coerceAtMost(chars.size)).joinToString("")
}

File: ios/App/Chat/message/Typewriter.swift
// Typewriter — pure time→reveal engine porting webui use-typewriter-buffer.ts.
// Rates from the shared SDK tokens (MobileData.Typewriter). Pure + testable; a
// TimelineView driver in the streaming bubble calls `typewriterTick` per frame.
import Foundation
import MobileData

struct TypewriterConfig {
    let baseRate: Double, minRate: Double, maxRate: Double, gapGain: Double
    let sentencePause: Double, paragraphPause: Double  // seconds
    static let `default` = TypewriterConfig(
        baseRate: Double(MobileData.Typewriter.shared.baseRate),
        minRate: Double(MobileData.Typewriter.shared.minRate),
        maxRate: Double(MobileData.Typewriter.shared.maxRate),
        gapGain: MobileData.Typewriter.shared.gapGain,
        sentencePause: Double(MobileData.Typewriter.shared.sentencePauseMs) / 1000.0,
        paragraphPause: Double(MobileData.Typewriter.shared.paragraphPauseMs) / 1000.0
    )
}

struct TypewriterState {
    var visibleCount: Int = 0
    var pauseUntil: Double = 0   // absolute seconds; reveal holds while now < pauseUntil
}

/// Advance `visibleCount` toward `target.count` for elapsed `dt` at clock `now`.
/// Stops at the first sentence/paragraph boundary within the computed advance window,
/// so the pause is never skipped when a tick would jump over a boundary.
func typewriterTick(_ s: TypewriterState, target: [Character], streamComplete: Bool,
                    dt: Double, now: Double, cfg: TypewriterConfig = .default) -> TypewriterState {
    var st = s
    let n = target.count
    if st.visibleCount >= n { return st }
    if now < st.pauseUntil { return st }
    let gap = n - st.visibleCount
    let raw = streamComplete ? cfg.maxRate : cfg.baseRate * (1 + Double(gap) * cfg.gapGain)
    let rate = min(cfg.maxRate, max(cfg.minRate, raw))
    let advance = max(1, Int(rate * dt))
    let candidate = min(n, st.visibleCount + advance)
    // Scan for the first boundary in (currentVisible, candidate]; stop there so
    // fast ticks never skip past a sentence/paragraph boundary.
    let stop = firstBoundaryStop(target, from: st.visibleCount + 1, through: candidate, cfg: cfg)
    st.visibleCount = stop
    if let pause = boundaryPause(target, upto: st.visibleCount, cfg: cfg) {
        st.pauseUntil = now + pause
    }
    return st
}

/// Returns the index (1-based count) of the first boundary char in [from, through],
/// or `through` if none found.
private func firstBoundaryStop(_ target: [Character], from: Int, through: Int,
                                cfg: TypewriterConfig) -> Int {
    for i in from...through {
        if boundaryPause(target, upto: i, cfg: cfg) != nil { return i }
    }
    return through
}

/// Semantic pause after the just-revealed char: paragraph (\n\n) > sentence (.!?).
private func boundaryPause(_ target: [Character], upto: Int, cfg: TypewriterConfig) -> Double? {
    guard upto >= 1, upto <= target.count else { return nil }
    let last = target[upto - 1]
    if last == "\n", upto >= 2, target[upto - 2] == "\n" { return cfg.paragraphPause }
    if last == "." || last == "!" || last == "?" { return cfg.sentencePause }
    return nil
}
